### PR TITLE
Train lora

### PR DESCRIPTION
**Pull request by khai-meetkai**
+ update code for training lora
+ merge fix_schema to this branch: also fix generating schema for array

### CODE REVIEW

File: ds_config/other_zero3.json
Comment (musab-mk) [inline ds_config/other_zero3.json:1]: can you put these configs under train folder

File: ds_config/other_zero3.json [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+        "loss_scale_window": 1000,
+        "initial_scale_power": 16,
+        "hysteresis": 2,
+        "min_loss_scale": 1
+    },
+    "bf16": {
+        "enabled": "auto"
+    },
+    "optimizer": {
+        "type": "AdamW",
```
Comment (musab-mk) [inline ds_config/other_zero3.json:14]: We can use paged adam if we need better memory consumption

File: functionary/train/train_lora.py [hunk 2]
```diff
@@ -0,0 +1,396 @@
...
+
+import torch
+import torch.distributed
+from torch.nn import CrossEntropyLoss
+import sys
+import os
+
+os.environ["WANDB_LOG_MODEL"] = "all"
+
+sys.path.append(os.path.join(os.path.dirname(__file__), "../.."))
```
Comment (musab-mk) [inline functionary/train/train_lora.py:16]: Why is this needed ? We should use module running like "python3 -m functionary.train.XXX" and it shouldnt be needed

File: functionary/train/train_lora.py [hunk 3]
```diff
@@ -0,0 +1,396 @@
...
+@dataclass
+class TrainingArguments(transformers.TrainingArguments):
+    cache_dir: Optional[str] = field(default=None)
+    model_max_length: int = field(
+        default=4096,
+        metadata={"help": "Maximum sequence length. Sequences will be right padded (and possibly truncated)."},
+    )
+
+
+def safe_save_model_for_hf_trainer(trainer: transformers.Trainer, output_dir: str):
```
Comment (musab-mk) [inline functionary/train/train_lora.py:57]: Let's use the new saving API that we used in other train.py

File: functionary/train/train_lora.py [hunk 4]
```diff
@@ -0,0 +1,396 @@
...
+    lora_r: int = 16
+    lora_alpha: int = 64
+    lora_dropout: float = 0.1
+    lora_target_modules: str = "all"  # all for all linear; "q_proj v_proj"
+    lora_weight_path: str = ""
+    lora_bias: str = "none"
+    q_lora: bool = False
+
+
+def initialize_tokenizer(
```
Comment (musab-mk) [inline functionary/train/train_lora.py:77]: Can you use the same tokenizer initialization instead of duplicating here

File: functionary/train/train_lora.py [hunk 5]
```diff
@@ -0,0 +1,396 @@
...
+
+    if training_args.gradient_checkpointing:
+        model.enable_input_require_grads()
+
+    model.config.use_cache = False
+    print_trainable_parameters(model)
+    return model
+
+
+def maybe_zero_3(param):
```
Comment (musab-mk) [inline functionary/train/train_lora.py:234]: can you add docstring or comments what this does

File: functionary/train/train_lora.py [hunk 6]
```diff
@@ -0,0 +1,396 @@
...
+
+    print_rank0("tokenizer.model_max_length: ", tokenizer.model_max_length)
+
+    model = prepare_model_for_training(model, training_args, lora_args)
+
+    assert data_args.train_data_path is not None, "Please provide a training data file."
+
+    with open(data_args.train_data_path, "r") as file:
+        raw_train_data = [json.loads(line) for line in file]
+    random.shuffle(raw_train_data)
```
Comment (musab-mk) [inline functionary/train/train_lora.py:298]: we removed shuffling from the training script.

File: functionary/train/train_lora.py [hunk 7]
```diff
@@ -0,0 +1,396 @@
...
+
+    if data_args.eval_data_path is not None:
+        with open(data_args.eval_data_path, "r") as file:
+            raw_eval_data = [json.loads(line) for line in file]
+        random.shuffle(raw_eval_data)
```
Comment (musab-mk) [inline functionary/train/train_lora.py:303]: same

File: functionary/train/train_lora.py [hunk 8]
```diff
@@ -0,0 +1,396 @@
...
+
+    train_dataset = CustomDataset(raw_train_data[:100], tokenizer)
```
Comment (musab-mk) [inline functionary/train/train_lora.py:305]: 100 ?

File: functionary/train/train_lora.py [hunk 9]
```diff
@@ -0,0 +1,396 @@
...
+
+    # if torch.distributed.get_rank() == 0:
+    #    print(f"Training Data Loaded: #{len(raw_train_data)}")
+
+    if training_args.do_eval:
+        eval_dataset = CustomDataset(raw_eval_data[:5], tokenizer)
```
Comment (musab-mk) [inline functionary/train/train_lora.py:311]: 5?

**Review by musab-mk — CHANGES_REQUESTED** [review]

File: functionary/schema.py [hunk 10]
```diff
@@ -33,26 +288,9 @@ def generate_schema_from_functions(functions: List[Function], namespace="functio
...
-                # Param Type
-                param_type = param.get("type", "any")
-                if param_type == "integer":
-                    param_type = "number"
-                if "enum" in param:
-                    param_type = " | ".join([f'"{v}"' for v in param["enum"]])
-                schema += f": {param_type},\n"
-
-            schema += "}) => any;\n\n"
+            info_lines = get_parameter_typescript(parameters.get("properties"), required_params, 0)
```
Comment (musab-mk) [inline functionary/schema.py:291]: can we rename this variable much more appropriately

File: functionary/schema.py [hunk 11]
```diff
@@ -9,7 +9,262 @@
...
+
+    Args:
+        properties (_type_): properties in parameters
+        required_params (_type_): List of required parameters
+        depth (int, optional): the depth of params (nested level). Defaults to 0.
+
+    Returns:
+        _type_: list of lines containing information about all parameters
+    """
+    info_lines = []
```
Comment (musab-mk) [inline functionary/schema.py:222]: can we rename this to be much more appropriate.

**Review by musab-mk — CHANGES_REQUESTED** [review]

File: ds_config/other_zero3.json
Comment (khai-meetkai) [reply ds_config/other_zero3.json:1]: Sure

File: ds_config/other_zero3.json (hunk 1 above)
Comment (khai-meetkai) [reply ds_config/other_zero3.json:14]: Not sure if paged adam works with deepspeed, I will test and confirm

File: functionary/train/train_lora.py (hunk 3 above)
Comment (khai-meetkai) [reply functionary/train/train_lora.py:57]: Actually I wrote the code from train.py  so include this function, but we don't need to use it, so I will delete

File: functionary/train/train_lora.py (hunk 4 above)
Comment (khai-meetkai) [reply functionary/train/train_lora.py:77]: what do you mean by using the same tokenizer initialization?

File: functionary/train/train_lora.py (hunk 5 above)
Comment (khai-meetkai) [reply functionary/train/train_lora.py:234]: Actually I copy this from train_lora fastchat, I will make a comment there
[maybe_zero_3](https://github.com/lm-sys/FastChat/blob/main/fastchat/train/train_lora.py#L68)

File: functionary/train/train_lora.py (hunk 6 above)
Comment (khai-meetkai) [reply functionary/train/train_lora.py:298]: Sure

File: functionary/train/train_lora.py (hunk 7 above)
Comment (khai-meetkai) [reply functionary/train/train_lora.py:303]: Sure

File: functionary/train/train_lora.py (hunk 8 above)
Comment (khai-meetkai) [reply functionary/train/train_lora.py:305]: Oh sorry, I only take first 100 items to make the training fast, forgot to remove it

File: functionary/train/train_lora.py (hunk 9 above)
Comment (khai-meetkai) [reply functionary/train/train_lora.py:311]: I took 5 items so the training would be fast, will fix it now

File: functionary/schema.py [hunk 12]
```diff
@@ -9,7 +9,262 @@
...
+    """
+    param_type = "any"
+    if "type" in param:
+        raw_param_type = param["type"]
+        if type(raw_param_type) is list:
+            param_type = " | ".join(raw_param_type)
+        else:
+            param_type = raw_param_type
+
+    else:  # in many cases, the json schema contains: oneOf instead of "type"
```
[the remark this answers is not in the document]
Comment (khai-meetkai) [reply functionary/schema.py:43]: Yes, I converted "oneOf" from json schema to type1|type2 
Screen Shot 2023-10-18 at 11 41 18

File: functionary/schema.py [hunk 13]
```diff
@@ -9,7 +9,262 @@
...
+        item_info = get_array_typescript(None, items_info, depth + 1)
+        if param_name is None:
+            return f"{item_info}[]"
+        return f"{offset}{param_name}: {item_info.strip()}[]"
+
+    else:
+        if "enum" in items_info:
+            item_type = get_enum_option_str(items_info["enum"])
+            if param_name is None:
+                return f"({item_type})[]"
```
[the remark this answers is not in the document]
Comment (khai-meetkai) [reply functionary/schema.py:200]: Oh this is for handling array so we need to add [], because array is quite complex so I wrote a separate function for handling array:
`def get_array_typescript(param_name: Optional[str], param_dic: dict, depth: int = 0) -> str:`

so if x is an array of enum:
Json schema:
`"x": {"type": "array", "items": {"type": "string", "enum": ["option1", "option2", "option3"]}}
`Type script:
`x: ("option1"|"option2"|"option3")[]`

File: functionary/schema.py [hunk 14]
```diff
@@ -9,7 +9,262 @@
...
+
+    if "default" in param:
+        default_value = param["default"]
+        if param_type == "string":
+            default_value = f'"{default_value}"'  # if string --> add ""
+        info_list.append(f"Default value={default_value}.")
+
+    format_param = get_format_param(param)
+    if format_param is not None:
+        info_list.append("The format is: " + format_param)
```
[the remark this answers is not in the document]
Comment (khai-meetkai) [reply functionary/schema.py:100]: Sure

File: functionary/schema.py (hunk 10 above)
Comment (khai-meetkai) [reply functionary/schema.py:291]: Each line is a line of code in typescript, 
so I change to: tp_lines ?

File: functionary/schema.py (hunk 11 above)
Comment (khai-meetkai) [reply functionary/schema.py:222]: renamed to  info_lines  -> tp_lines

File: functionary/schema.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (khai-meetkai) [reply functionary/schema.py:200]: You can take a look at: https://github.com/MeetKai/functionary/blob/train_lora/tests/test_schema.py
to see the output of converting json schema to typescript schema

File: functionary/train/train_lora.py (hunk 2 above)
Comment (khai-meetkai) [reply functionary/train/train_lora.py:16]: Oh this is because we are using deepspeed so the command should be:
deepspeed functionary/train/train_lora.py
If we run: 
deepspeed -m functionary.train.train_lora

File: functionary/train/train_lora.py (hunk 4 above)
Comment (musab-mk) [reply functionary/train/train_lora.py:68]: I meant, we already have initialize_tokenizer in the train.py. So we can import it instead

File: functionary/schema.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (khai-meetkai) [reply functionary/schema.py:200]: class A (Enum, str):
 option1 = "1"
 option2 = "2"
 items = List[A]

File: functionary/schema.py [hunk 15]
```diff
@@ -9,7 +9,262 @@
...
+
+    if "default" in param:
+        default_value = param["default"]
+        if param_type == "string":
+            default_value = f'"{default_value}"'  # if string --> add ""
+        info_list.append(f"Default={default_value}.")
+
+    format_param = get_format_param(param)
+    if format_param is not None:
+        info_list.append("Format:" + format_param)
```
Comment (musab-mk) [inline functionary/schema.py:100]: Let's use either "=" or ":"

File: functionary/schema.py [hunk 16]
```diff
@@ -9,7 +9,262 @@
...
+
+    for field, field_name in [
+        ("maximum", "Maximum"),
+        ("minimum", "Minimum"),
+        ("maxLength", "Maximum length"),
+        ("minLength", "Minimum length"),
+    ]:
+        if field in param:
+            info_list.append(f"{field_name}=" + str(param[field]) + ".")
```
Comment (musab-mk) [inline functionary/schema.py:109]: "." is not necessary.

**Review by musab-mk — CHANGES_REQUESTED** [review]

File: functionary/prompt.py [hunk 17]
```diff
@@ -37,12 +37,27 @@ def from_message(cls, message: Dict) -> EndToken:
...
                 return EndToken.function_call
             else:
                 return EndToken.assistant
 
 
+class BeginToken(str, Enum):
+    function = "<|BEGIN_OF_FUNCTION_CALL|>"
+    
+
+def get_added_tokens() -> List[str]:
```
Comment (musab-mk) [inline functionary/prompt.py:50]: can we rename this to be get_additional_tokens or get_special_tokens

File: functionary/prompt.py [hunk 18]
```diff
@@ -37,12 +37,27 @@ def from_message(cls, message: Dict) -> EndToken:
...
             return EndToken.function
         else:  # role = assistant
-            if "function_call" in message:
+            if "function_call" in message and message["function_call"] is not None:
                 return EndToken.function_call
             else:
                 return EndToken.assistant
 
 
+class BeginToken(str, Enum):
```
Comment (musab-mk) [inline functionary/prompt.py:46]: It should be named as "StartToken" as the exact opposite of end is "start"

File: functionary/prompt.py [hunk 19]
```diff
@@ -37,12 +37,27 @@ def from_message(cls, message: Dict) -> EndToken:
...
+    function = "<|BEGIN_OF_FUNCTION_CALL|>"
```
Comment (musab-mk) [inline functionary/prompt.py:47]: Lets also rename this to be start. and all related code too

File: functionary/prompt.py [hunk 20]
```diff
@@ -53,36 +68,33 @@ def get_text_from_message(message: Dict) -> str:
...
     stop_token = EndToken.from_message(message).value
-    content = message.get("content", "")
-
-    if content is not None:
-        content = f"{content}{stop_token}"
+    content = message.get("content", None)
 
     if message["role"] == "system":
-        text = "system:\n{content}\n".format(content=content)
+        text = f"system:\n{content}{stop_token}\n"
```
Comment (musab-mk) [inline functionary/prompt.py:74]: can we use consistent naming. as we used EndToken, lets use end_token instead of stop_token

File: tests/prompt_test.txt [hunk 21]
```diff
@@ -17,13 +17,14 @@ assistant:
...
-assistant to=get_car_price:
+assistant:
+I need to get the price of of the car Song<|BEGIN_OF_FUNCTION_CALL|>to=get_car_price:
 {
   "car_name": "Song"
 }<|END_OF_FUNCTION_CALL|>
 function name=get_car_price:
 {'price': {'price': '$25000'}}<|END_OF_FUNCTION_RESULT|>
-assistant to=get_car_price:
+assistant <|BEGIN_OF_FUNCTION_CALL|>to=get_car_price:
```
Comment (musab-mk) [inline tests/prompt_test.txt:27]: we should be consistent about "assistant:"
also lets remove the "to="

**Review by musab-mk — CHANGES_REQUESTED** [review]

File: functionary/prompt.py (hunk 18 above)
Comment (khai-meetkai) [reply functionary/prompt.py:46]: sure

File: functionary/prompt.py (hunk 19 above)
Comment (khai-meetkai) [reply functionary/prompt.py:47]: Sure

File: functionary/prompt.py (hunk 17 above)
Comment (khai-meetkai) [reply functionary/prompt.py:50]: sure

File: functionary/prompt.py (hunk 20 above)
Comment (khai-meetkai) [reply functionary/prompt.py:74]: sure

File: functionary/schema.py (hunk 15 above)
Comment (khai-meetkai) [reply functionary/schema.py:100]: Ok

File: functionary/schema.py (hunk 16 above)
Comment (khai-meetkai) [reply functionary/schema.py:109]: OK

File: tests/prompt_test.txt (hunk 21 above)
Comment (khai-meetkai) [reply tests/prompt_test.txt:27]: Sure
